File: packages/sdk-js/src/scope.ts
import type { TelemetryBreadcrumb, User, ScopeData } from './types';

/**
 * Scope manages contextual data for events
 * Includes user info, tags, extra data, and breadcrumbs
 */
export class Scope implements ScopeData {
  user?: User;
  tags: Record<string, string> = {};
  extra: Record<string, unknown> = {};
  breadcrumbs: TelemetryBreadcrumb[] = [];
  fingerprint?: string[];

  private maxBreadcrumbs: number;

  constructor(maxBreadcrumbs: number = 100) {
    this.maxBreadcrumbs = maxBreadcrumbs;
  }

  /**
   * Set user context
   */
  setUser(user: User | null): void {
    this.user = user ?? undefined;
  }

  /**
   * Set a tag
   */
  setTag(key: string, value: string): void {
    this.tags[key] = value;
  }

  /**
   * Set multiple tags
   */
  setTags(tags: Record<string, string>): void {
    Object.assign(this.tags, tags);
  }

  /**
   * Set extra context
   */
  setExtra(key: string, value: unknown): void {
    this.extra[key] = value;
  }

  /**
   * Set multiple extra values
   */
  setExtras(extras: Record<string, unknown>): void {
    Object.assign(this.extra, extras);
  }

  /**
   * Set fingerprint for grouping
   */
  setFingerprint(fingerprint: string[]): void {
    this.fingerprint = fingerprint;
  }

  /**
   * Add a breadcrumb (ring buffer - oldest removed when full)
   */
  addBreadcrumb(breadcrumb: Omit<TelemetryBreadcrumb, 'timestamp'>): void {
    const crumb: TelemetryBreadcrumb = {
      ...breadcrumb,
      timestamp: new Date().toISOString(),
    };

    this.breadcrumbs.push(crumb);

    // Maintain ring buffer size
    if (this.breadcrumbs.length > this.maxBreadcrumbs) {
      this.breadcrumbs.shift();
    }
  }

  /**
   * Clear breadcrumbs
   */
  clearBreadcrumbs(): void {
    this.breadcrumbs = [];
  }

  /**
   * Clear all scope data
   */
  clear(): void {
    this.user = undefined;
    this.tags = {};
    this.extra = {};
    this.breadcrumbs = [];
    this.fingerprint = undefined;
  }

  /**
   * Clone scope for isolation
   */
  clone(): Scope {
    const clone = new Scope(this.maxBreadcrumbs);
    clone.user = this.user ? { ...this.user } : undefined;
    clone.tags = { ...this.tags };
    clone.extra = { ...this.extra };
    clone.breadcrumbs = [...this.breadcrumbs];
    clone.fingerprint = this.fingerprint ? [...this.fingerprint] : undefined;
    return clone;
  }

  /**
   * Apply scope data to an error context
   */
  applyToContext(): Omit<ScopeData, 'breadcrumbs'> & { breadcrumbs: TelemetryBreadcrumb[] } {
    return {
      user: this.user,
      tags: { ...this.tags },
      extra: { ...this.extra },
      breadcrumbs: [...this.breadcrumbs],
      fingerprint: this.fingerprint,
    };
  }
}

// ---------------------------------------------------------------------------
// Scope ID storage abstraction (mirrors context.ts ContextStorage pattern)
// ---------------------------------------------------------------------------

interface ScopeIdStorage {
  get(): string | null;
  set(id: string | null): void;
  run<T>(id: string, fn: () => T): T;
  runAsync<T>(id: string, fn: () => Promise<T>): Promise<T>;
}

class SimpleScopeIdStorage implements ScopeIdStorage {
  private current: string | null = null;

  get(): string | null {
    return this.current;
  }

  set(id: string | null): void {
    this.current = id;
  }

  run<T>(id: string, fn: () => T): T {
    const prev = this.current;
    this.current = id;
    try {
      return fn();
    } finally {
      this.current = prev;
    }
  }

  async runAsync<T>(id: string, fn: () => Promise<T>): Promise<T> {
    const prev = this.current;
    this.current = id;
    try {
      return await fn();
    } finally {
      this.current = prev;
    }
  }
}

class AsyncLocalScopeIdStorage implements ScopeIdStorage {
  private als: import('node:async_hooks').AsyncLocalStorage<string | null>;

  constructor(als: import('node:async_hooks').AsyncLocalStorage<string | null>) {
    this.als = als;
  }

  get(): string | null {
    return this.als.getStore() ?? null;
  }

  set(id: string | null): void {
    this.als.enterWith(id);
  }

  run<T>(id: string, fn: () => T): T {
    return this.als.run(id, fn);
  }

  runAsync<T>(id: string, fn: () => Promise<T>): Promise<T> {
    return this.als.run(id, fn);
  }
}

function createScopeIdStorage(): ScopeIdStorage {
  if (
    typeof process !== 'undefined' &&
    typeof process.versions?.node !== 'undefined'
  ) {
    try {
      // eslint-disable-next-line @typescript-eslint/no-require-imports
      const { AsyncLocalStorage } = require('node:async_hooks');
      return new AsyncLocalScopeIdStorage(new AsyncLocalStorage());
    } catch {
      // Fallback
    }
  }
  return new SimpleScopeIdStorage();
}

/**
 * Scope manager with AsyncLocalStorage support for context propagation
 */
export class ScopeManager {
  private globalScope: Scope;
  private isolationScopes: Map<string, Scope> = new Map();
  private scopeIdStorage: ScopeIdStorage = createScopeIdStorage();

  constructor(maxBreadcrumbs: number = 100) {
    this.globalScope = new Scope(maxBreadcrumbs);
  }

  /**
   * Get the current active scope
   */
  getCurrentScope(): Scope {
    const currentScopeId = this.scopeIdStorage.get();
    if (currentScopeId) {
      return this.isolationScopes.get(currentScopeId) ?? this.globalScope;
    }
    return this.globalScope;
  }

  /**
   * Get the global scope
   */
  getGlobalScope(): Scope {
    return this.globalScope;
  }

  /**
   * Run a function with an isolated scope
   */
  withScope<T>(callback: (scope: Scope) => T): T {
    const scopeId = crypto.randomUUID();
    const scope = this.globalScope.clone();
    this.isolationScopes.set(scopeId, scope);

    try {
      return this.scopeIdStorage.run(scopeId, () => callback(scope));
    } finally {
      this.isolationScopes.delete(scopeId);
    }
  }

  /**
   * Run an async function with an isolated scope
   */
  async withScopeAsync<T>(callback: (scope: Scope) => Promise<T>): Promise<T> {
    const scopeId = crypto.randomUUID();
    const scope = this.globalScope.clone();
    this.isolationScopes.set(scopeId, scope);

    try {
      return await this.scopeIdStorage.runAsync(scopeId, () => callback(scope));
    } finally {
      this.isolationScopes.delete(scopeId);
    }
  }
}

File: packages/sdk-python/syntra/scope.py
"""Scope management for Syntra SDK."""

from __future__ import annotations

import contextvars
from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any, Callable, TypeVar

from syntra.types import Breadcrumb, BreadcrumbLevel, BreadcrumbType, User

T = TypeVar("T")

# Context variable for scope isolation
_current_scope: contextvars.ContextVar[Scope | None] = contextvars.ContextVar(
    "syntra_scope", default=None
)


@dataclass
class Scope:
    """
    Scope manages contextual data for events.

    Includes user info, tags, extra data, and breadcrumbs.
    """

    user: User | None = None
    tags: dict[str, str] = field(default_factory=dict)
    extra: dict[str, Any] = field(default_factory=dict)
    breadcrumbs: list[Breadcrumb] = field(default_factory=list)
    fingerprint: list[str] | None = None
    _max_breadcrumbs: int = field(default=100, repr=False)

    def set_user(self, user: User | None) -> None:
        """Set user context."""
        self.user = user

    def set_tag(self, key: str, value: str) -> None:
        """Set a tag."""
        self.tags[key] = value

    def set_tags(self, tags: dict[str, str]) -> None:
        """Set multiple tags."""
        self.tags.update(tags)

    def set_extra(self, key: str, value: Any) -> None:
        """Set extra context."""
        self.extra[key] = value

    def set_extras(self, extras: dict[str, Any]) -> None:
        """Set multiple extra values."""
        self.extra.update(extras)

    def set_fingerprint(self, fingerprint: list[str]) -> None:
        """Set fingerprint for grouping."""
        self.fingerprint = fingerprint

    def add_breadcrumb(
        self,
        type: BreadcrumbType = BreadcrumbType.DEFAULT,
        category: str = "default",
        message: str | None = None,
        data: dict[str, Any] | None = None,
        level: BreadcrumbLevel = BreadcrumbLevel.INFO,
    ) -> None:
        """Add a breadcrumb (ring buffer - oldest removed when full)."""
        breadcrumb = Breadcrumb(
            type=type,
            category=category,
            message=message,
            data=data,
            level=level,
            timestamp=datetime.now(timezone.utc).isoformat() + "Z",
        )
        self.breadcrumbs.append(breadcrumb)

        # Maintain ring buffer size
        while len(self.breadcrumbs) > self._max_breadcrumbs:
            self.breadcrumbs.pop(0)

    def clear_breadcrumbs(self) -> None:
        """Clear all breadcrumbs."""
        self.breadcrumbs = []

    def clear(self) -> None:
        """Clear all scope data."""
        self.user = None
        self.tags = {}
        self.extra = {}
        self.breadcrumbs = []
        self.fingerprint = None

    def clone(self) -> Scope:
        """Clone scope for isolation."""
        return Scope(
            user=dict(self.user) if self.user else None,  # type: ignore
            tags=dict(self.tags),
            extra=dict(self.extra),
            breadcrumbs=list(self.breadcrumbs),
            fingerprint=list(self.fingerprint) if self.fingerprint else None,
            _max_breadcrumbs=self._max_breadcrumbs,
        )


class ScopeManager:
    """Manages scopes including global and isolation scopes."""

    def __init__(self, max_breadcrumbs: int = 100) -> None:
        self._global_scope = Scope(_max_breadcrumbs=max_breadcrumbs)
        self._max_breadcrumbs = max_breadcrumbs

    def get_current_scope(self) -> Scope:
        """Get the current active scope."""
        scope = _current_scope.get()
        return scope if scope is not None else self._global_scope

    def get_global_scope(self) -> Scope:
        """Get the global scope."""
        return self._global_scope


# Global scope manager instance
_scope_manager: ScopeManager | None = None


def get_scope_manager() -> ScopeManager:
    """Get the global scope manager."""
    global _scope_manager
    if _scope_manager is None:
        _scope_manager = ScopeManager()
    return _scope_manager


def set_scope_manager(manager: ScopeManager) -> None:
    """Set the global scope manager."""
    global _scope_manager
    _scope_manager = manager


def get_current_scope() -> Scope:
    """Get the current scope."""
    return get_scope_manager().get_current_scope()


def with_scope(callback: Callable[[Scope], T]) -> T:
    """
    Run a function with an isolated scope.

    Args:
        callback: Function to run with the isolated scope

    Returns:
        The return value of the callback

    Example:
        def process():
            with_scope(lambda scope: (
                scope.set_tag("operation", "process"),
                do_something()
            ))
    """
    manager = get_scope_manager()
    scope = manager.get_current_scope().clone()

    token = _current_scope.set(scope)
    try:
        return callback(scope)
    finally:
        _current_scope.reset(token)


async def with_scope_async(callback: Callable[[Scope], Any]) -> Any:
    """
    Run an async function with an isolated scope.

    Args:
        callback: Async function to run with the isolated scope

    Returns:
        The return value of the callback
    """
    manager = get_scope_manager()
    scope = manager.get_current_scope().clone()

    token = _current_scope.set(scope)
    try:
        return await callback(scope)
    finally:
        _current_scope.reset(token)
